File: project/swig/ruby/metadata.rb
#!/usr/bin/env ruby

require 'librets'
include Librets

def dump_system(metadata)
  system = metadata.system
  puts "System ID: " + system.system_id
  puts "Description: " + system.system_description
  puts "Comment: " + system.comments
end

def dump_all_resources(metadata)
  puts
  metadata.GetAllResources.each do |resource|
    dump_all_classes(metadata, resource)
  end
  metadata.GetAllResources.each do |resource|
    dump_all_lookups(metadata, resource)
  end
end

def dump_all_classes(metadata, resource)
  resource_name = resource.resource_id;
  metadata.GetAllClasses(resource_name).each do |aClass|
    puts "Resource name: " +  resource_name + " [" + 
      resource.standard_name + "]"
    puts "Class name: " + aClass.class_name + " [" +
      aClass.GetStandardName + "]"
    dump_all_tables(metadata, aClass)
    puts
  end
end

def dump_all_tables(metadata, aClass)
  metadata.all_tables(aClass).each do |table|
    puts "Table name: " + table.system_name + " [" + table.standard_name +
      "]"
  end
end

def dump_all_lookups(metadata, resource)
  resource_name = resource.resource_id();
  metadata.all_lookups(resource_name).each do |lookup|
    puts "Resource name: " +  resource_name + " [" + 
      resource.standard_name + "]"
    puts "Lookup name: " + lookup.lookup_name + " (" +
      lookup.visible_name + ")"
    dump_all_lookup_types(metadata, lookup)
    puts
  end
end

def dump_all_lookup_types(metadata, lookup)
  metadata.all_lookup_types(lookup).each do |lookup_type|
    puts "Lookup value: " + lookup_type.value + " (" +
      lookup_type.short_value + ", " +
      lookup_type.long_value + ")"
  end
end

session = RetsSession.new("http://demo.crt.realtors.org:6103/rets/login")
if (ARGV.length == 1) && (ARGV[0] == "full")
  session.incremental_metadata = false
end

if !session.login("Joe", "Schmoe")
  puts "Invalid login"
  exit 1
end

metadata = session.metadata
dump_system(metadata)
dump_all_resources(metadata)

session.logout()

File: project/swig/perl/metadata.pl
#!/usr/bin/env perl

use lib "blib/lib", "blib/arch";
use strict;
use librets;
use Data::Dumper;

# Here we define some helper functions
sub dumpAllTables
{
    my $metadata = shift;
    my $class = shift;

    my $tables = $metadata->GetAllTables($class);
    foreach my $table (@$tables)
    {
        print "Table name: " . $table->GetSystemName() . " [" .
            $table->GetStandardName() . "]\n";
        print "\tType: " . $table->GetDataType() . "\n";
        print "\tUnique: " . $table->IsUnique() . "\n";
        print "\tMax Length: " . $table->GetMaximumLength() . "\n";
    }
}

sub dumpAllClasses
{
    my $metadata = shift;
    my $resource = shift;

    my $classes = $metadata->GetAllClasses($resource->GetResourceID());
    foreach my $class (@$classes)
    {
        print "Class name: " . $class->GetClassName() . " [" .
            $class->GetStandardName() . "]\n";
        dumpAllTables($metadata, $class);
    }
}

sub dumpAllLookupTypes
{
    my $metadata = shift;
    my $lookup = shift;

    my $lookupTypes = $metadata->GetAllLookupTypes($lookup);
    foreach my $lt (@$lookupTypes)
    {
        print "Lookup value: " . $lt->GetValue() . " (" .
            $lt->GetShortValue() . ", " . $lt->GetLongValue() . ")\n";
    }
}

sub dumpAllLookups
{
    my $metadata = shift;
    my $resource = shift;

    my $lookups = $metadata->GetAllLookups($resource->GetResourceID());
    foreach my $lookup (@$lookups)
    {
        print "Lookup name: " . $lookup->GetLookupName() . " (" .
            $lookup->GetVisibleName() . ")\n";
        dumpAllLookupTypes($metadata, $lookup);
        print "\n";
    }
}
    

my $session = new librets::RetsSession(
    "http://demo.crt.realtors.org:6103/rets/login");

if (!$session->Login("Joe", "Schmoe"))
{
    print "Invalid login\n";
    exit 1;
}

my $metadata = $session->GetMetadata();
my $system = $metadata->GetSystem();
print "System ID: " . $system->GetSystemID() . "\n";
print "Desription: " . $system->GetSystemDescription() . "\n";
print "Comment : " . $system->GetComments() . "\n\n";

my $resources = $metadata->GetAllResources();
foreach my $resource (@$resources)
{
    print "Resource name: " . $resource->GetResourceID() . " [" .
        $resource->GetStandardName() . "]\n";
    dumpAllClasses($metadata, $resource);
    dumpAllLookups($metadata, $resource);
    print "\n";
    
}

$session->Logout();
